community/collections/src/main/java/org/neo4j/helpers/collection/CastingIterator.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.helpers.collection;
+
+import java.util.Iterator;
+
+/**
+ * An iterator which filters another iterator, only letting items with certain
+ * criteria pass through. All iteration/filtering is done lazily.
+ *
+ * @param <T> the type of items in the iteration.
+ */
+public class CastingIterator<T extends A, A> extends PrefetchingIterator<T>
```

Comment:
The `FilteringIterator` is almost the same, but it does not cast to the subtype, which is really nice for the calling code. Fixed javadoc (which was not correct) to clarify the difference.